TW-235: remove wildcards from Visibility docs (#1919)
temporalio/documentation: docs-src/concepts/what-is-a-list-filter.md

---

A List Filter is the SQL-like string that is provided as the parameter to an [Advanced Visibility](/concepts/what-is-advanced-visibility) List API.
List Filter [Search Attribute](/concepts/what-is-a-search-attribute) names are case sensitive. List Filter applies to a single [Namespace](/concepts/what-is-a-namespace).

A List Filter that uses a time range has a resolution of 1 ns on Elasticsearch 7.
The range of a List Filter timestamp (StartTime, CloseTime, ExecutionTime) cannot exceed 9223372036854775807 (that is, maxInt64 - 1001).

### Supported operators

A List Filter contains [Search Attribute](/concepts/what-is-a-search-attribute) names, Search Attribute values, and the following supported operators:

- **=, !=, >, >=, <, <=**
- **AND, OR, ()**
- **BETWEEN ... AND**
- **IN**
- **ORDER BY**

The `=` operator works like **CONTAINS** to find Workflows with Search Attributes that contain a specific word.

> The **ORDER BY** operator is supported only with [Advanced Visibility](/clusters/how-to-integrate-elasticsearch-into-a-temporal-cluster).

Author: Which is in Advanced Visibility

> The **ORDER BY** operator is supported only with [Advanced Visibility](/clusters/how-to-integrate-elasticsearch-into-a-temporal-cluster).

Reviewer: I'm confused by the terminology. Are we keeping the "Advanced Visibility" meaning as visibility with Elasticsearch? What about the upcoming changes introducing advanced visibility features in SQL DB?

> The **ORDER BY** operator is supported only with [Advanced Visibility](/clusters/how-to-integrate-elasticsearch-into-a-temporal-cluster).

Reviewer 2: Yeah; maybe this is better put as
>  The `ORDER BY` operator is supported only with Elasticsearch as the Visibility store.

IIRC @DJSanti has already updated this in a recent commit

### Partial string match

If a `Description` Search Attribute of `Text` type is set to a phrase, List Filters containing words within that phrase will return the Workflow.
Custom Search Attributes of `Text` type cannot be used in **ORDER BY** clauses.

- `Description="quick"` or `Description="fox"` would both return a Workflow with the Description Search Attribute set to "The quick brown fox jumps over the lazy dog."
- `Description="qui"` or `Description="laz"` would not return the Workflow.
  [Elasticsearch's tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-standard-tokenizer.html) is configured to return complete words as tokens.
- To use a partial string match (i.e. specific word), create a [custom Search Attribute](/app-dev-context/observability#custom-search-attributes) of type `Text` and set it value to a phrase that contains that word.

### Efficient API usage

> An Advanced List Filter API may take longer to respond if it is retrieving a large number of Workflow Executions (over 10 million, for instance).

Reviewer 3: I don't know where 10M came from. Server has a max page size setting. By default it is 10000. Most likely original phrase meant total number of workflows in visibility index. But in this case `Count` won't help.

Author: Good to know. This was in the original page.

> Use the `CountWorkflow` API to efficiently count the number of [Workflow Executions](/concepts/what-is-a-workflow-execution).

Author: Okay. I'll comment out the line for now.

Paginate the results with the `ListWorkflow` API by using the page token to retrieve the next page; continue until the page token is `null`/`nil`.

#### List Filter examples

The following is a List Filter set with [`tctl`](/tctl-v1/workflow/list):

```

> WorkflowType = "main.YourWorkflowDefinition" and ExecutionStatus != "Running" and (StartTime > "2021-06-07T16:46:34.236-08:00" or CloseTime > "2021-06-07T16:46:34-08:00") order by StartTime desc

Reviewer: Can we remove the `order by` part? Might cause confusion as in SQL is not supported.

```

> When used, a list of Workflows that meet the following conditions are returned, ordered by `StartTime` in descending order:

Reviewer: Remove maybe if it's removed above as well?

---

A List Filter is the SQL-like string that is provided as the parameter to an [Advanced Visibility](/concepts/what-is-advanced-visibility) List API.
List Filter [Search Attribute](/concepts/what-is-a-search-attribute) names are case sensitive, and each List Filter is scoped by a single [Namespace](/concepts/what-is-a-namespace).

A List Filter that uses a time range has a resolution of 1 ns on Elasticsearch 7+ and 1 ms for SQL.

### Supported operators

A List Filter contains [Search Attribute](/concepts/what-is-a-search-attribute) names, Search Attribute values, and the following supported operators:

- **=, !=, >, >=, <, <=**
- **AND, OR, ()**
- **BETWEEN ... AND**
- **IN**
- **ORDER BY**

> The **ORDER BY** operator is supported only for Elasticsearch used as [Advanced Visibility](/clusters/how-to-integrate-elasticsearch-into-a-temporal-cluster).

Reviewer 4: ```suggestion
The **ORDER BY** operator is supported only when Elasticsearch is used as the Visibility store.
```

- **AND, OR, ()**
- **BETWEEN ... AND**
- **IN**
- **ORDER BY**

> The **ORDER BY** operator is supported only when Elasticsearch is used as the Visibility store.

Reviewer 4: ```suggestion
The **ORDER BY** operator is supported only when Elasticsearch is used as the Visibility store.
Additionally, custom Search Attributes of the `Text` type cannot be used in **ORDER BY** clauses.
```

Reviewer 4: I just think it reads better and makes more sense to group the "ORDER BY" information - don't you?

> For example, if you have a Search Attribute `Description` with the value of "The quick brown fox jumps over the lazy dog", searching for `Description=quick` or `Description=fox` will successfully return the Workflow.

Reviewer: ```suggestion
For example, if you have a Search Attribute `Description` of `Text` type with the value of "The quick brown fox jumps over the lazy dog", searching for `Description=quick` or `Description=fox` will successfully return the Workflow.
```

The `=` operator works like **CONTAINS** to find Workflows with Search Attributes that contain a specific word.

<!-- note: advanced vis features will be supported in SQL upon the release of v1.20.-->

For example, if you have a Search Attribute `Description` of `Text` type with the value of "The quick brown fox jumps over the lazy dog", searching for `Description=quick` or `Description=fox` will successfully return the Workflow.
However, partial word searches such as `Description=qui` or `Description=laz` will not return the Workflow.
This is because [Elasticsearch's tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-standard-tokenizer.html) is configured to return complete words as tokens.

### Efficient API usage

An Advanced List Filter API may take longer to respond if it is retrieving a large number of Workflow Executions (over 10,000).

> <!-- Use the `CountWorkflow` API to efficiently count the number of [Workflow Executions](/concepts/what-is-a-workflow-execution). -->

Reviewer: We have delayed v1.20 release, and we were able to include `CountWorkflow` in it. This can be uncommented now.

Author: Until 1.2 is out, this is staying commented.

Author: We can make another update in the future for all 1.2 updates. This PR was focused on correcting existing information that could mislead users at present.

  - **BETWEEN ... AND**
  - **IN**
  - **ORDER BY**

>   The last operator (**ORDER BY**) is only supported with Elasticsearch as a Visibility store.

Reviewer 5: ```suggestion
  The last operator (**ORDER BY**) is supported only with Elasticsearch as a Visibility store.
```

  Custom Search Attributes of `Text` type cannot be used in **ORDER BY** clauses.

- To search for a specific word, set a [custom Search Attribute](/app-dev-context/observability#custom-search-attributes) of type `Text` to a phrase that contains that word.

>   Any command run with that Search Attribute returns any Workflows that contain that word.

Reviewer 3: I don't understand this last sentence.

> - To search for a specific word, set a [custom Search Attribute](/app-dev-context/observability#custom-search-attributes) of type `Text` to a phrase that contains that word.
>   Any command run with that Search Attribute returns any Workflows that contain that word.

Reviewer 3: ```suggestion
- To use a partial string match (i.e. specific word), create a [custom Search Attribute](/app-dev-context/observability#custom-search-attributes) of type `Text` and set it value to a phrase that contains that word. Use equal operator `=` in List Filters to find Workflows which have Search Attribute that contain that word.
```

And I do believe that good example will really help here. Something like:
===
If `Description` Search Attribute of type `Text` is set to "The quick brown fox jumps over the lazy dog" then List Filters like `Description="quick"` or `Description="fox"` will return this workflow. Note that it works only for full word match. List Filters like `Description="qui"` or `Description="laz"` won't return this workflow. Read more about standard Elasticsearch tokenizer at https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-standard-tokenizer.html
===

> - To search for a specific word, [set a custom Search Attribute of type `Text`](/app-dev-context/observability#custom-search-attributes) to a phrase that contains that word.
>   Any command run with that Search Attribute set to it will return any Workflows that contain that word.

Reviewer 5: ```suggestion
- To search for a specific word, set a [custom Search Attribute](/app-dev-context/observability#custom-search-attributes) of type `Text` to a phrase that contains that word.
  Any command run with that Search Attribute returns any Workflows that contain that word.
```
Two reasons for the change in the link text:
- For SEO, link text should match the target as closely as possible.
- Recent style change: Link text should not contain inline code (monospace).

  - **BETWEEN ... AND**
  - **IN**
  - **ORDER BY**

>   The last operator (**ORDER BY**) is supported only with Elasticsearch as the Visibility store.

Reviewer 6: I would maybe instead of "Elasticsearch as Visibility store" put "Advanced Visibility" and link to docs page that describes advanced visibility, think there is a page on this right?

Reviewer 4: @DJSanti - I like the sentence, but let's definitely link to `/clusters/how-to-integrate-elasticsearch-into-a-temporal-cluster`

@tsurdilo Can we assume then that `ORDER BY` is available with Cloud Visibility?

  Custom Search Attributes of `Text` type cannot be used in **ORDER BY** clauses.

> - To use a partial string match (i.e. specific word), create a [custom Search Attribute](/app-dev-context/observability#custom-search-attributes) of type `Text` and set it value to a phrase that contains that word.

Reviewer 6: I would double check this with @alexshtin if this is by default whole-word search or partial as in https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-regexp-query.html#_standard_operators is allowed too.

Reviewer 6: ah there is sample below. okok sry

> - The `=` operator works like **CONTAINS** to find Workflows with Search Attributes that contain a specific word.

Reviewer 4: @DJSanti I would reorder these bullets and talk about how the `=` operator works first - then back that up with the use case of partial string matches - and the additional examples that Alex has - 

And make sure the information is properly nested.. right now - these bullets are at the same level - so someone would maybe get confused about whether the information was linear.

Author: Idea: it's reordered but not as a bulleted list.

Author: Idea: reorder these bullet points, but not as a list. Instead, write out the page some while keeping this information.

  - **ORDER BY**
  - **LIKE**

> - To search for a specific token, set a custom Search Attribute of type `Text` that has the value of the token(s) separated by spaces or punctuation. In the example provided below, the Query will run after setting the custom Search Attribute to "type1":

Reviewer 5: ```suggestion
- To search for a specific token, set a custom Search Attribute of type `Text` that has the value of the token.
  To search for multiple tokens, separate them by spaces or punctuation.
  In the following example, the Query runs after setting the custom Search Attribute to `type1`:
```

  ```

>   tctl workflow list --query "mySearchAttribute = "\type1\""

Reviewer 5: ```suggestion
  tctl workflow list --query "mySearchAttribute = \"type1\""
```

  - **ORDER BY**

>   - **LIKE**

Reviewer 4: Now would be a good time to identify that LIKE and ORDER BY are only supported with Elasticsearch as a Visibility store.

@aarohib ?

Author: Added a comment about this. Waiting for @aarohib 's input

Reviewer 3: Let's remove `LIKE` and `%` from everywhere in the doc. It doesn't work the way most people expect it to work and everything which it does can be achieved with simple `=` operator.

- To search for a specific token, set a custom Search Attribute of type `Text` that has the value of the token.
  To search for multiple tokens, separate them by spaces or punctuation.

>   In the following example, the Query runs after setting the custom Search Attribute to `type1`:

Reviewer 4: We should link to examples in each lang + the CLI tool

Author: I'll link to the examples and tctl. I'll be sure to change it to CLI when that's all finally settled

Author: Links added. One for tctl-next and one to the Observability section that has code examples in tabs

  - **ORDER BY**
  - **LIKE**

  The last two operators, **ORDER BY** and **LIKE**, are only supported with Elasticsearch as a Visibility store.

> - To search for a specific token, [set a custom Search Attribute of type `Text`](/app-dev-context/observability#custom-search-attributes) that has the value of the token.

Reviewer 3: I think you rather need to explain what token is (in pop up of something) or replace it with more common word "word".

Reviewer 3: I read this sentence few times and I don't get it. Set is set, search is search. First, you set value of Search Attribute of type `Text` to something like "The quick brown fox jumps over the lazy dog" and then search for specific word "fox".

  To search for multiple tokens, separate them by spaces or punctuation.

>   In the following example for [`tctl-next`](/tctl-next/workflow/list), the Query runs after setting the custom Search Attribute to `type1`:

Reviewer 3: ```suggestion
  In the following example for [`tctl-next`](/tctl-next/workflow#list), the Query runs after setting the custom Search Attribute to `type1`:
```
?

  ```

>   tctl workflow list --query "mySearchAttribute = \"type1\""

Reviewer 3: This example is irrelevant here. If you want to describe how search attributes of type `Text` works you better set it to some phrase first and then search for specific word. I also would like to avoid names like "mySearchAttribute" and values like "type1". Of course, it is random text but by making it more meaningful you make example better. This why I used `ProductId`.

Author: There are examples below. Going to remove this bit.

  ```

>   Any Workflow Ids that contain this token will be returned.

Reviewer 3: ```suggestion
  All Workflows that contain this token will be returned.
```
visibility queries returns entire workflow record, not Id only.
